### PR TITLE
Sagaize actions/login code

### PR DESCRIPTION
**Pull request by MarcoPolo**
@keybase/react-hackers 

The transformation is mostly straightforward from the old login incoming map to the new saga map.

The new saga map defines sagas for each incoming message. When we receive a message we start the corresponding saga. The saga navAppends a new screen and takes on some actions.

Screens no longer rely on closures that are passed in. They do the same as every other component. They just call dispatch on an action from an actionCreator.

Still needs a lot of testing. But I've tested the paperkey flow, and adding from device flow. On desktop and mobile

### CODE REVIEW

**Comment by cjb** [comment]
Doing some testing: Settings->Sign out no longer works.

**Comment by cjb** [comment]
After doing a new provision and clicking on a paper key, I see "mode is null!, can't setCodePageMode. Bailing" in console.  The provision worked though.

**Comment by MarcoPolo** [comment]
This was probably a problem with the old code too, I added another check to prevent it from doing something likely wrong. (RE: after starting a new provision...)

**Comment by keybase-ci-visdiff** [comment]
The commits 93eb36978590...7eff6712970f introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Code_Page-Show_Text_Mobile** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1491845770463-93eb36978590-7eff6712970f-linux/Code_Page-Show_Text_Mobile-93eb36978590.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1491845770463-93eb36978590-7eff6712970f-linux/Code_Page-Show_Text_Mobile-7eff6712970f.png)  
   Code_Page-Show_Text_Mobile comparison
</summary>

File: shared/actions/login/saga.js [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+  }
+}
+
+// TODO add waiting handlers
+// TODO sagaize
+const waitingForResponse = (waiting: boolean) : TypedAction<'login:waitingForResponse', boolean, void> => (
+  {payload: waiting, type: Constants.waitingForResponse}
+)
+
+const makeWaitingHandler = (dispatch: Dispatch): {waitingHandler: (waiting: boolean) => void} => (
```
Comment (chromakode) [inline shared/actions/login/saga.js]: This was a bit hard to read at first -- hard to re-read it a few times to wrap my head around where the type definition was.

File: shared/actions/login/saga.js [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+  yield put(Creators.loginDone())
+  yield put(loadDevices())
+  yield put(bootstrap())
+}
+
+function * addNewDeviceSaga ({payload: {role}}: DeviceConstants.AddNewDevice) {
+  yield call(initalizeMyCodeStateForAddingADevice)
+
+  const onBackSaga = function * (response) {
+    yield put(loadDevices())
```
Comment (chromakode) [inline shared/actions/login/saga.js]: :speech_balloon: Bummer that we have to explicitly call this here and in other places. We should move this into the connected component init or some kind of route entry handler.

File: shared/actions/login/saga.js [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+    'finished': finishedSaga,
+  }
+
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(reloginSagas))
+  const chanMap = Types.loginLoginProvisionedDeviceRpcChannelMap(
+    channelConfig,
+    {param: {noPassphrasePrompt: false, username: usernameOrEmail}},
+  )
+
+  yield _.map(reloginSagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatest(c, saga), chanMap, methodName))
```
Comment (chromakode) [inline shared/actions/login/saga.js]: :speech_balloon: A thought: do the channel map / sagas get cleaned up effectively by GC when they are finished?

File: shared/actions/login/saga.js [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+          type: Constants.actionForgotPasswordDone,
+        })
+      }
+    },
+  }
+
+  const email = yield select(state => state.login.forgotPasswordEmailAddress)
+  const channelConfig = Saga.singleFixedChannelConfig(Object.keys(sagas))
+  const chanMap = Types.loginRecoverAccountFromEmailAddressRpcChannelMap(channelConfig, {param: {email}})
+  yield _.map(sagas, (saga, methodName) => Saga.effectOnChannelMap(c => Saga.safeTakeLatest(c, saga), chanMap, methodName))
```
Comment (chromakode) [inline shared/actions/login/saga.js]: Seems like you do this operation a lot... would it make sense to move into a helper (which might help name/describe what this is doing better)?

File: shared/login/login/routes.js [hunk 5]
```diff
@@ -16,6 +16,7 @@ const recursiveLazyRoutes = I.Seq({
   login: {
     component: Login,
   },
+  // TODO: update this component
```
Comment (chromakode) [inline shared/login/login/routes.js]: Is this still valid? If so, can you please elaborate on what needs to be done?

File: shared/actions/login/saga.js (hunk 1 above)
Comment (MarcoPolo) [inline shared/actions/login/saga.js]: yeah this was old code that still needs to be updated

File: shared/actions/login/saga.js (hunk 3 above)
Comment (MarcoPolo) [inline shared/actions/login/saga.js]: Nope. making a ticket: DESKTOP-3508

**Comment by keybase-ci-visdiff** [comment]
The commits 93eb36978590...c1cd24eebfc6 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Code_Page-Show_Text_Mobile** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1492668088436-93eb36978590-c1cd24eebfc6-linux/Code_Page-Show_Text_Mobile-93eb36978590.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1492668088436-93eb36978590-c1cd24eebfc6-linux/Code_Page-Show_Text_Mobile-c1cd24eebfc6.png)  
   Code_Page-Show_Text_Mobile comparison
</summary>

**Pull request merged**